src/vm.rs

```diff
@@ -853,6 +914,50 @@ impl VM {
                 }
             }
 
+            Op::Iter => {
+                let ty: Type = match Type::from(self.pop()) {
+                    Type::List(e) => {
+                        e.as_ref().clone()
+                    }
+                    Type::Tuple(v) => {
+                        let set: HashSet<_> = v.into_iter().collect();
+                        match set.len() {
+                            0 => Type::Unknown,
+                            1 => set.into_iter().next().unwrap(),
+                            _ => Type::Union(set),
+                        }
+                    }
+                    Type::Set(v) => {
+                        v.as_ref().clone()
+                    }
+                    Type::Dict(k, v) => {
+                        Type::Tuple(vec![k.as_ref().clone(), v.as_ref().clone()])
+                    }
+                    i => {
+                        error!(
+                            self,
+                            ErrorKind::TypeError(Op::Iter, vec![i]),
+                            "Cannot convert to iterator"
+                        );
+                    }
+                };
+                self.push(Value::Iter(ty, Rc::new(RefCell::new(|| None))));
+            }
+
+            Op::JmpNext(_) => {
+                if let Some(Value::Iter(ty, _)) = self.stack.last() {
+                    let v = Value::from(ty);
+                    self.push(v);
+                } else {
+                    self.push(Value::Nil);
+                    error!(
+                        self,
+                        ErrorKind::InvalidProgram,
+                        "Can only 'next' iterators"
```

Comment:
I'm guessing you mean to have different error messages so you can grep more easily and not have to care about if the error is when checking or running, but I think the solution is actually to report the errors differently. I've been thinking of something like this:

```rust
pub enum Error {
    CompileError {
        kind: ErrorKind,
        file: Option<PathBuf>,
        line: Option<usize>,
        message: Option<String>,
    },
    TypecheckError {
        kind: ErrorKind,
        file: PathBuf,
        line: usize,
        message: Option<String>,
    },
    RuntimeError {
        kind: ErrorKind,
        file: PathBuf,
        line: usize,
        message: Option<String>,
    },
    BincodeError,
}
```

We'd have to change some things in our test files, but that may be for the better since we'd have to specify *when* an error should be found. We don't want to miss when a type error used to be found during type checking but is suddenly only found in the runtime.

Reply:
It's fine for this PR however, since we can't (easily) see a difference currently.